File: app/src/main/java/com/example/agriventure/ui/market/FarmerMarketFragment.java
package com.example.agriventure.ui.market;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.GridLayout;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.agriventure.R;
import com.example.agriventure.data.adapter.ProductsAdapter;
import com.example.agriventure.data.models.Produce;
import com.example.agriventure.ui.BaseFragment;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class FarmerMarketFragment extends BaseFragment {

    private List<Produce> myProductList;
    private RecyclerView myProductsRv, allProductsRv;
    private MaterialTextView marketEmptyText, myProduceTitle, allProduceTitle;
    private MaterialButton buttonAddProduce;
    private ProgressBar mProgressBar;
    private AppCompatImageView emptyImage;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_farmer_market, container, false);
        myProductsRv = view.findViewById(R.id.user_market_products);
        allProductsRv = view.findViewById(R.id.all_market_products);

        marketEmptyText = view.findViewById(R.id.market_empty);
        myProduceTitle = view.findViewById(R.id.user_market_title);
        allProduceTitle = view.findViewById(R.id.other_market_title);

        buttonAddProduce = view.findViewById(R.id.btn_add_produce);

        mProgressBar = view.findViewById(R.id.data_loading);

        emptyImage = view.findViewById(R.id.img_market_empty);

        myProductList = new ArrayList<>();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        DatabaseReference mDatabase = FirebaseDatabase.getInstance().getReference().child("produce");

        mDatabase.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                myProductList.clear();
                for (DataSnapshot dataSnapshot : snapshot.getChildren()){
                    Produce produce = dataSnapshot.getValue(Produce.class);
                    myProductList.add(produce);
                }
                updateView();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

        buttonAddProduce.setOnClickListener(v -> {
            Navigation.findNavController(getView()).navigate(R.id.action_navigation_farmer_market_to_navigation_add_produce);
        });
    }

    private void updateView() {
        if (myProductList.size()>0){
            myProduceTitle.setVisibility(View.VISIBLE);
            buttonAddProduce.setVisibility(View.VISIBLE);
            setUpMyProducts(myProductList);
        }else{
            buttonAddProduce.setVisibility(View.VISIBLE);
            marketEmptyText.setVisibility(View.VISIBLE);
            emptyImage.setVisibility(View.VISIBLE);
            mProgressBar.setVisibility(View.GONE);
        }
    }

    private void setUpMyProducts(List<Produce> products) {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(activity, RecyclerView.VERTICAL, false);
       // GridLayoutManager gridLayoutManager = new GridLayoutManager(activity, 2);
        ProductsAdapter productsAdapter = new ProductsAdapter(activity, products, produce -> {
            Bundle bundle = new Bundle();
            bundle.putParcelable("produce", produce);
            Navigation.findNavController(getView()).navigate(R.id.action_navigation_farmer_market_to_navigation_produce_detail, bundle);
        });
        myProductsRv.setAdapter(productsAdapter);
        myProductsRv.setLayoutManager(linearLayoutManager);
        myProductsRv.setVisibility(View.VISIBLE);
        allProductsRv.setVisibility(View.VISIBLE);
        mProgressBar.setVisibility(View.GONE);
    }
}

File: app/src/main/java/com/example/agriventure/ui/market/FarmerMarketFragment.kt
package com.example.agriventure.ui.market

import android.content.Context
import com.example.agriventure.ui.BaseFragment
import com.example.agriventure.data.models.Produce
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textview.MaterialTextView
import com.google.android.material.button.MaterialButton
import android.widget.ProgressBar
import androidx.appcompat.widget.AppCompatImageView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.navigation.Navigation
import com.example.agriventure.R
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.agriventure.data.adapter.ProductsAdapter
import com.example.agriventure.data.adapter.ProductsAdapter.ProduceClickListener
import com.example.agriventure.databinding.FragmentFarmerMarketBinding
import com.example.agriventure.util.Constants
import java.util.ArrayList

class FarmerMarketFragment : BaseFragment() {
    private lateinit var myProductList: MutableList<Produce>
    private lateinit var allProduceList: MutableList<Produce>

    lateinit var binding : FragmentFarmerMarketBinding
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentFarmerMarketBinding.inflate(inflater, container, false)
        myProductList = ArrayList()
        allProduceList = ArrayList()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        val userName = activity?.getPreferences(Context.MODE_PRIVATE)?.getString(Constants.userName,"")
        val businessName = activity?.getPreferences(Context.MODE_PRIVATE)?.getString(Constants.farmerBusinessName,"")
        binding.welcomeTitle.text = "Welcome, "+userName

        val mDatabase = FirebaseDatabase.getInstance().reference.child("produce")
        mDatabase.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                myProductList.clear()
                for (dataSnapshot in snapshot.children) {
                    val produce = dataSnapshot.getValue(Produce::class.java)
                    produce?.let {
                        if (produce.seller_name == businessName){
                            myProductList.add(it)
                        }else{
                            allProduceList.add(it)
                        }
                    }
                }
                updateView()
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        binding.btnAddProduce.setOnClickListener { v: View? ->
            Navigation.findNavController(
                requireView()
            ).navigate(R.id.action_navigation_farmer_market_to_navigation_add_produce)
        }
    }

    private fun updateView() {
        binding.btnAddProduce.visibility = View.VISIBLE
        var validate = allProduceList.isEmpty() || myProductList.isEmpty()
        if (allProduceList.isEmpty() and myProductList.isEmpty()){
            binding.btnAddProduce.visibility = View.VISIBLE
            binding.marketEmpty.visibility = View.VISIBLE
            binding.imgMarketEmpty.visibility = View.VISIBLE
            binding.dataLoading.visibility = View.GONE
        }else{
            setUpMyProducts(myProductList)
            setUpAllProducts(allProduceList)
        }
    }

    private fun setUpMyProducts(products: List<Produce?>?) {
        if (products!=null && products.isNotEmpty()){
            val linearLayoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
            val myProductsAdapter = ProductsAdapter(activity, products) { produce: Produce? ->
                val bundle = Bundle()
                bundle.putParcelable("produce", produce)
                Navigation.findNavController(requireView())
                    .navigate(R.id.action_navigation_farmer_market_to_navigation_produce_detail, bundle)
            }
            binding.userMarketProducts.adapter = myProductsAdapter
            binding.userMarketProducts.layoutManager = linearLayoutManager
            binding.userMarketProducts.visibility = View.VISIBLE
            binding.userMarketTitle.visibility = View.VISIBLE
            binding.dataLoading.visibility = View.GONE
        }
    }

    private fun setUpAllProducts(products: List<Produce?>?) {
        if (products!=null && products.isNotEmpty()) {
            val linearLayoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
            val allProductsAdapter = ProductsAdapter(activity, products) { produce: Produce? -> }
            binding.allMarketProducts.adapter = allProductsAdapter
            binding.allMarketProducts.layoutManager = linearLayoutManager
            binding.allMarketProducts.visibility = View.VISIBLE
            binding.allMarketTitle.visibility = View.VISIBLE
            binding.dataLoading.visibility = View.GONE
        }
    }
}